File: python/TwoPointers/ContainerWithMostWater.py
"""
11. Container With Most Water
Difficulty: Medium
Topics: Array, Two Pointers, Greedy
"""
def maxArea(height: list[int]) -> int:
    n = len(height)
    max_area = 0
    left, right = 0, n-1  # two pointers
    
    while left < right: 
        # Area(rectangle) = height * width
        # Height
        r_height = min(height[left], height[right])
        # Width
        r_width = right - left
        # Max Area
        max_area = max(max_area, r_height * r_width)

        # We move the pointer with the lower value (1st left pointer < 1st right pointer in the example)
        if height[left] < height[right]:
            left += 1
        else:
            right -= 1

    return max_area

File: java/TwoPointers/ContainerWithMostWater.java
package java.TwoPointers;

public class ContainerWithMostWater {
    /*
     * 11. Container With Most Water
     * Difficulty: Medium
     * Topics: Array, Two Pointers, Greedy
     */
    public static int maxArea(int[] height) {
        int N = height.length;
        int maxArea = 0;
        int left = 0; int right = N-1;

        while (left < right) {
            int vertical = Math.min(height[left], height[right]);
            int horizontal = right - left;
            maxArea = Math.max(maxArea, vertical * horizontal);

            if (height[left] < height[right]) {
                left++;
            } else {
                right--;
            }
        }

        return maxArea;
    }
}
